### PR TITLE
[SPARKR] [SPARK-10981] SparkR Join improvements

### PR DESCRIPTION
**Pull request by mfliu**
I was having issues with collect() and orderBy() in Spark 1.5.0 so I used the DataFrame.R file and test_sparkSQL.R file from the Spark 1.5.1 download. I only modified the join() function in DataFrame.R to include "full", "fullouter", "left", "right", and "leftsemi" and added corresponding test cases in the test for join() and merge() in test_sparkSQL.R file.
Pull request because I filed this JIRA bug report:
https://issues.apache.org/jira/browse/SPARK-10981

### CODE REVIEW

**Comment by shivaram** [comment]
Thanks for PR @mfliu  - Could you format the PR title as `[SPARKR] [SPARK-10981] SparkR Join improvements` ? More instructions are at https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark#ContributingtoSpark-PreparingtoContributeCodeChanges

**Comment by shivaram** [comment]
Jenkins, ok to test

**Comment by shivaram** [comment]
cc @sun-rui

**Comment by mfliu** [comment]
The errors in the log file don't seem to be related to the changes I made? They are primarily in PythonRDD.scala:
java.net.SocketException: Socket is closed

org.apache.spark.SparkException: Python worker exited unexpectedly (crashed)
	at org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRDD.scala:203)
	at org.apache.spark.api.python.PythonRunner$$anon$1.<init>(PythonRDD.scala:207)
	at org.apache.spark.api.python.PythonRunner.compute(PythonRDD.scala:125)
	at org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:70)

Caused by: java.io.EOFException
	at java.io.DataInputStream.readInt(DataInputStream.java:392)
	at org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRDD.scala:139)

Errors seem to be from PythonRDD.scala, but I made no changes to that file, and I'm not sure how changing R code affects that interface?

Can someone else take a look? @shivaram, @sun-rui

**Comment by shivaram** [comment]
@mfliu I think thats just a flaky test, but irrespective of that your changes should be against the current `master` branch. Right now it looks like there are a lot more lines in the diff because of the change being against 1.5.1 ? For fixing this in branch-1.5, we will first merge with master and then during merge we can backport it.

File: R/pkg/R/DataFrame.R [hunk 1]
```diff
@@ -1314,50 +1273,21 @@ setClassUnion("characterOrColumn", c("character", "Column"))
...
-
-            sdf <- callJMethod(x@sdf, "sort", jcols)
-            dataFrame(sdf)
-          })
-
-#' @rdname arrange
-#' @export
-setMethod("arrange",
-          signature(x = "DataFrame", col = "character"),
-          function(x, col, ..., decreasing = FALSE) {
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: It looks like this is undoing a recent PR, could you check?

File: R/pkg/R/DataFrame.R [hunk 2]
```diff
@@ -1854,30 +1784,36 @@ setMethod("fillna",
...
+#'
+#' @rdname statfunctions
+#' @name crosstab
+#' @export
+#' @examples
+#' \dontrun{
+#' df <- jsonFile(sqlCtx, "/path/to/file.json")
+#' ct = crosstab(df, "title", "gender")
 #' }
-setMethod("as.data.frame",
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: also for this from a recent PR

**Comment by mfliu** [comment]
@felixcheung Yes, you are correct. The arrange function was different. I pulled again and changed those files and it is working on my machine. Can you test again?

**Comment by mfliu** [comment]
Sorry, had a typo in one of my unit tests. It now passes the run-tests.sh on my machine. Can you test again?

File: R/pkg/R/DataFrame.R [hunk 3]
```diff
@@ -1414,9 +1414,10 @@ setMethod("where",
 #' @param x A Spark DataFrame
 #' @param y A Spark DataFrame
 #' @param joinExpr (Optional) The expression used to perform the join. joinExpr must be a
-#' Column expression. If joinExpr is omitted, join() wil perform a Cartesian join
+#' Column expression. If joinExpr is omitted, join() will perform a Cartesian join
 #' @param joinType The type of join to perform. The following join types are available:
-#' 'inner', 'outer', 'left_outer', 'right_outer', 'semijoin'. The default joinType is "inner".
+#' 'inner', 'outer', 'full', 'fullouter', leftouter', 'left',
```
Comment (shivaram) [inline R/pkg/R/DataFrame.R]: any reason we should change `right_outer` to `rightouter` ? It'll break code that used to work with previous versions ?

**Comment by mfliu** [comment]
That change was made based on the comment on the JIRA report:
https://issues.apache.org/jira/browse/SPARK-10981

In the PR, please:
1. Support all join types defined in sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/joinTypes.scala (You can remove the "_" char from the currently supported join types in SparkR)
2. Add test cases for missing join types including "leftsemi"

Perhaps I misunderstood?

File: R/pkg/R/DataFrame.R (hunk 3 above)
Comment (mfliu) [inline R/pkg/R/DataFrame.R]: That change was made based on the comment on the JIRA report:
https://issues.apache.org/jira/browse/SPARK-10981

In the PR, please:
1. Support all join types defined in sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/joinTypes.scala (You can remove the "_" char from the currently supported join types in SparkR)
2. Add test cases for missing join types including "leftsemi"

Perhaps I misunderstood?

Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: +1 breaking API changes, IMO we really need to come up on some policy on that

Comment (mfliu) [inline R/pkg/R/DataFrame.R]: Would it make sense to add "right_outer" and "left_outer" along with "rightouter" and "leftouter"?

Comment (sun-rui) [inline R/pkg/R/DataFrame.R]: yeah, API compatibility is a concern. So we can make R code consistent with the scala version at https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/joinTypes.scala#L21

That is, replace the "_" char in the join type string with empty.

Comment (shivaram) [inline R/pkg/R/DataFrame.R]: Yeah, so lets support both right_outer and rightouter. That way we don't break backwards compatibility. One simple way to do this as @sun-rui said is to just replace all "_"s in the join string with "" using `gsub` or something like that.

**Comment by shivaram** [comment]
@mfliu Just FYI, you can check the lint-r tests locally by running the script `dev/lint-r`

**Comment by shivaram** [comment]
Jenkins, retest this please

**Comment by felixcheung** [comment]
looks good

**Comment by sun-rui** [comment]
LGTM

**Comment by shivaram** [comment]
Thanks @mfliu - LGTM. Merging this

**Pull request closed without merging**